src/value.cpp

```diff
@@ -498,12 +500,30 @@ MemRef::getDimsAndLayoutAndElemTy(
 
 pair<z3::expr, z3::expr> MemRef::load(const vector<z3::expr> &indices) {
   z3::expr idx = to1DIdxWithLayout(indices);
-  return m->load(bid, offset + idx);
+  auto [expr, success] = m->load(bid, offset + idx);
+
+  // check whether indices are  inbound.
+  for (int i = 0; i < indices.size(); i ++)
+    success = success && z3::ult(indices[i], getDim(i));
+
```

Comment:
This may contradicts with https://github.com/aqjune/mlir-tv/pull/32#discussion_r669383881.
But in mlir documents (https://mlir.llvm.org/docs/Dialects/Builtin/#index-map), the index space is always non-negative and integral. So it is not allowed to access with negative Index.

And I think it is fair to mark out-of-bounds access as UB, as out-of-bounds pointer access is UB.
With out this, below `var_buffer_cast.src.mlir`- `var_buffer_cast.tgt.mlir` pair results in "Source is more defined than target" because tensor already checks out-of-bounds access and mark it as UB

Reply:
Hmm.. it seems like difficult issue.
Maybe this dimension wise out-of-bounds check should be relaxed.

```
int arr[64][64];
return arr[4][100]; // 4*64 + 100 < 64*64, so it is well defined?
```

Reply:
Although above expression is admitted in C semantic, dimension-wise index bounds are strictly checked in MLIR. 
`mlir::boundCheckLoadOrStoreOp` in `mlir/lib/Analysis/Utils.cpp` may relate with this.
So we need this bound checks in well-defineness.